RBD: add snap delete function

iPraveenParihar opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #

- rbd: add rbdSnap.Delete() function
This function deletes rbd snap and rbd image
backing k8s snapshot.
The same function is used for deleting
individual snapshots in group snapshot.

**Checklist:**

* [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [x] Reviewed the developer guide on [Submitting a Pull Request](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#development-workflow)
* [ ] [Pending release notes](https://github.com/ceph/ceph-csi/blob/devel/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next major release.
* [ ] Documentation has been updated, if necessary.
* [x] Unit tests have been added, if necessary.
* [x] Integration tests have been added, if necessary.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

iPraveenParihar committed:
e2e: validate images in trash for volumegroupsnapshot test

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.31

Madhu-1 commented on e2e/volumegroupsnapshot.go:
```diff
@@ -176,6 +176,13 @@ func (rvgs *rbdVolumeGroupSnapshot) ValidateResourcesForCreate(vgs *groupsnapapi
 	validateOmapCount(rvgs.framework, 0, rbdType, defaultRBDPool, volumesType)
+	validateOmapCount(rvgs.framework, 0, rbdType, defaultRBDPool, snapsType)
+	validateOmapCount(rvgs.framework, 0, rbdType, defaultRBDPool, groupSnapsType)
+
```
add a validation of no images in the pool itself. `validateRBDImageCount`

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.31

iPraveenParihar commented:
failed [ci logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e_k8s-1.31/runs/226/nodes/90/log/?start=0)
```
  I1204 09:58:16.710898       1 rbd_util.go:1506] ID: 52 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-8c1a0986-5ba1-4111-bcbb-5faac931bc14 rbd: snap rm replicapool/csi-snap-8c1a0986-5ba1-4111-bcbb-5faac931bc14@csi-snap-8c1a0986-5ba1-4111-bcbb-5faac931bc14 using mon rook-ceph-mon-a.rook-ceph.svc.cluster.local:6789
  I1204 09:58:16.713026       1 group_snapshot.go:98] ID: 51 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-5c1fbbd1-bc0a-415a-974e-a0cb1292261c resolved snapshot id "0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-728c142a-3231-428b-9429-884ff3ce40e6" to snapshot "replicapool/csi-snap-728c142a-3231-428b-9429-884ff3ce40e6@csi-snap-728c142a-3231-428b-9429-884ff3ce40e6"
  I1204 09:58:16.714216       1 omap.go:89] ID: 51 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-5c1fbbd1-bc0a-415a-974e-a0cb1292261c got omap values: (pool="replicapool", namespace="", name="csi.snap.8c1a0986-5ba1-4111-bcbb-5faac931bc14"): map[csi.groupid:0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-5c1fbbd1-bc0a-415a-974e-a0cb1292261c csi.imageid:22c224ead8e6 csi.imagename:csi-snap-8c1a0986-5ba1-4111-bcbb-5faac931bc14 csi.snapname:csi-vol-group-061ff308-e2a7-4ca8-b3a3-2ebeff75bea7-snap-1 csi.source:csi-vol-652e01f5-cf8e-4db8-9f99-a861be116889 csi.volume.owner:rbd-2737]
  I1204 09:58:16.741070       1 group_snapshot.go:98] ID: 51 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-5c1fbbd1-bc0a-415a-974e-a0cb1292261c resolved snapshot id "0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-8c1a0986-5ba1-4111-bcbb-5faac931bc14" to snapshot "replicapool/csi-snap-8c1a0986-5ba1-4111-bcbb-5faac931bc14@csi-snap-8c1a0986-5ba1-4111-bcbb-5faac931bc14"
  I1204 09:58:16.741107       1 group_snapshot.go:108] ID: 51 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-5c1fbbd1-bc0a-415a-974e-a0cb1292261c GetVolumeGroupSnapshot(0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-5c1fbbd1-bc0a-415a-974e-a0cb1292261c) returns {commonVolumeGroup:{id:0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-5c1fbbd1-bc0a-415a-974e-a0cb1292261c requestName:groupsnapshot-f47d779b-02c7-4513-8ff1-a912bd77fee8 name:csi-vol-group-5c1fbbd1-bc0a-415a-974e-a0cb1292261c creationTime:0xc000a8cd20 clusterID:63946011-0514-45b3-ac2c-f0a3ddf7967d objectUUID:5c1fbbd1-bc0a-415a-974e-a0cb1292261c credentials:0xc0007919c0 conn:<nil> ioctx:<nil> monitors:rook-ceph-mon-a.rook-ceph.svc.cluster.local:6789 pool:replicapool namespace: csiDriver:default journal:0xc000a1ef60} snapshots:[0xc000877200 0xc0001c4600 0xc000bc6000] snapshotsToFree:[0xc000877200 0xc0001c4600 0xc000bc6000]}
  I1204 09:58:16.741133       1 group_snapshot.go:231] ID: 51 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-5c1fbbd1-bc0a-415a-974e-a0cb1292261c deleting snapshot image "replicapool/csi-snap-9358d356-480a-458f-9728-e80a9f3ee8a5@csi-snap-9358d356-480a-458f-9728-e80a9f3ee8a5" for volume group snapshot "replicapool/csi-vol-group-5c1fbbd1-bc0a-415a-974e-a0cb1292261c"
  I1204 09:58:16.741186       1 rbd_util.go:1506] ID: 51 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-5c1fbbd1-bc0a-415a-974e-a0cb1292261c rbd: snap rm replicapool/csi-snap-9358d356-480a-458f-9728-e80a9f3ee8a5@csi-snap-9358d356-480a-458f-9728-e80a9f3ee8a5 using mon rook-ceph-mon-a.rook-ceph.svc.cluster.local:6789
  I1204 09:58:17.734862       1 rbd_util.go:653] ID: 51 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-5c1fbbd1-bc0a-415a-974e-a0cb1292261c rbd: delete csi-snap-9358d356-480a-458f-9728-e80a9f3ee8a5 using mon rook-ceph-mon-a.rook-ceph.svc.cluster.local:6789, pool replicapool
  I1204 09:58:17.735076       1 rbd_util.go:653] ID: 52 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-8c1a0986-5ba1-4111-bcbb-5faac931bc14 rbd: delete csi-snap-8c1a0986-5ba1-4111-bcbb-5faac931bc14 using mon rook-ceph-mon-a.rook-ceph.svc.cluster.local:6789, pool replicapool
  I1204 09:58:17.762222       1 rbd_util.go:695] ID: 51 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-5c1fbbd1-bc0a-415a-974e-a0cb1292261c rbd: adding task to remove image "replicapool/csi-snap-9358d356-480a-458f-9728-e80a9f3ee8a5" with id "22c23db22227" from trash
  I1204 09:58:17.764649       1 rbd_util.go:695] ID: 52 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-8c1a0986-5ba1-4111-bcbb-5faac931bc14 rbd: adding task to remove image "replicapool/csi-snap-8c1a0986-5ba1-4111-bcbb-5faac931bc14" with id "22c224ead8e6" from trash
  I1204 09:58:17.768331       1 rbd_util.go:719] ID: 51 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-5c1fbbd1-bc0a-415a-974e-a0cb1292261c rbd: successfully added task to move image "replicapool/csi-snap-9358d356-480a-458f-9728-e80a9f3ee8a5" with id "22c23db22227" to trash
  I1204 09:58:17.768369       1 group_snapshot.go:231] ID: 51 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-5c1fbbd1-bc0a-415a-974e-a0cb1292261c deleting snapshot image "replicapool/csi-snap-728c142a-3231-428b-9429-884ff3ce40e6@csi-snap-728c142a-3231-428b-9429-884ff3ce40e6" for volume group snapshot "replicapool/csi-vol-group-5c1fbbd1-bc0a-415a-974e-a0cb1292261c"
  I1204 09:58:17.768426       1 rbd_util.go:1506] ID: 51 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-5c1fbbd1-bc0a-415a-974e-a0cb1292261c rbd: snap rm replicapool/csi-snap-728c142a-3231-428b-9429-884ff3ce40e6@csi-snap-728c142a-3231-428b-9429-884ff3ce40e6 using mon rook-ceph-mon-a.rook-ceph.svc.cluster.local:6789
  I1204 09:58:17.774422       1 rbd_util.go:719] ID: 52 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-8c1a0986-5ba1-4111-bcbb-5faac931bc14 rbd: successfully added task to move image "replicapool/csi-snap-8c1a0986-5ba1-4111-bcbb-5faac931bc14" with id "22c224ead8e6" to trash
  I1204 09:58:17.780787       1 omap.go:126] ID: 52 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-8c1a0986-5ba1-4111-bcbb-5faac931bc14 removed omap keys (pool="replicapool", namespace="", name="csi.snaps.default"): [csi.snap.csi-vol-group-061ff308-e2a7-4ca8-b3a3-2ebeff75bea7-snap-1]
  I1204 09:58:17.781552       1 utils.go:273] ID: 52 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-8c1a0986-5ba1-4111-bcbb-5faac931bc14 GRPC response: {}
  I1204 09:58:18.751329       1 rbd_util.go:653] ID: 51 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-5c1fbbd1-bc0a-415a-974e-a0cb1292261c rbd: delete csi-snap-728c142a-3231-428b-9429-884ff3ce40e6 using mon rook-ceph-mon-a.rook-ceph.svc.cluster.local:6789, pool replicapool
  I1204 09:58:18.785256       1 rbd_util.go:695] ID: 51 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-5c1fbbd1-bc0a-415a-974e-a0cb1292261c rbd: adding task to remove image "replicapool/csi-snap-728c142a-3231-428b-9429-884ff3ce40e6" with id "22c2a95617af" from trash
  I1204 09:58:18.795433       1 rbd_util.go:719] ID: 51 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-5c1fbbd1-bc0a-415a-974e-a0cb1292261c rbd: successfully added task to move image "replicapool/csi-snap-728c142a-3231-428b-9429-884ff3ce40e6" with id "22c2a95617af" to trash
  I1204 09:58:18.795502       1 group_snapshot.go:231] ID: 51 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-5c1fbbd1-bc0a-415a-974e-a0cb1292261c deleting snapshot image "replicapool/csi-snap-8c1a0986-5ba1-4111-bcbb-5faac931bc14@csi-snap-8c1a0986-5ba1-4111-bcbb-5faac931bc14" for volume group snapshot "replicapool/csi-vol-group-5c1fbbd1-bc0a-415a-974e-a0cb1292261c"
  I1204 09:58:18.795560       1 rbd_util.go:1506] ID: 51 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-5c1fbbd1-bc0a-415a-974e-a0cb1292261c rbd: snap rm replicapool/csi-snap-8c1a0986-5ba1-4111-bcbb-5faac931bc14@csi-snap-8c1a0986-5ba1-4111-bcbb-5faac931bc14 using mon rook-ceph-mon-a.rook-ceph.svc.cluster.local:6789
  E1204 09:58:18.799257       1 snapshot.go:86] ID: 51 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-5c1fbbd1-bc0a-415a-974e-a0cb1292261c failed to delete snapshot "replicapool/csi-snap-8c1a0986-5ba1-4111-bcbb-5faac931bc14@csi-snap-8c1a0986-5ba1-4111-bcbb-5faac931bc14": Failed as image not found (internal RBD image not found: rbd: ret=-2, No such file or directory)
  E1204 09:58:18.799278       1 snapshot.go:181] ID: 51 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-5c1fbbd1-bc0a-415a-974e-a0cb1292261c failed to delete image: Failed as image not found (internal RBD image not found: rbd: ret=-2, No such file or directory)
  E1204 09:58:18.799344       1 utils.go:271] ID: 51 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-5c1fbbd1-bc0a-415a-974e-a0cb1292261c GRPC error: rpc error: code = Internal desc = failed to delete volume group snapshot "replicapool/csi-vol-group-5c1fbbd1-bc0a-415a-974e-a0cb1292261c": failed to delete snapshot "replicapool/csi-snap-8c1a0986-5ba1-4111-bcbb-5faac931bc14@csi-snap-8c1a0986-5ba1-4111-bcbb-5faac931bc14" as part of volume groups snapshot "replicapool/csi-vol-group-5c1fbbd1-bc0a-415a-974e-a0cb1292261c": Failed as image not found (internal RBD image not found: rbd: ret=-2, No such file or directory)
  ```
  
 `rbd: snap rm replicapool/csi-snap-8c1a0986-5ba1-4111-bcbb-5faac931bc14@csi-snap-8c1a0986-5ba1-4111-bcbb-5faac931bc14` was called twice and second time it failed.

```
ID: 52 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-8c1a0986-5ba1-4111-bcbb-5faac931bc14 GRPC call: /csi.v1.Controller/DeleteSnapshot
```
/DeleteSnapshot was called for replicapool/csi-snap-8c1a0986-5ba1-4111-bcbb-5faac931bc14

And this image was actually part of VolumeGroupSnapshot 
```
I1204 09:58:16.741070       1 group_snapshot.go:98] ID: 51 Req-ID: 0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-5c1fbbd1-bc0a-415a-974e-a0cb1292261c resolved snapshot id "0001-0024-63946011-0514-45b3-ac2c-f0a3ddf7967d-0000000000000004-8c1a0986-5ba1-4111-bcbb-5faac931bc14" to snapshot "replicapool/csi-snap-8c1a0986-5ba1-4111-bcbb-5faac931bc14@csi-snap-8c1a0986-5ba1-4111-bcbb-5faac931bc14"
```

When deleting VolumeGroupSnapshot,  snapshotter calls /DeleteSnapshot for the snapshot that is part of the VolumeGroupSnapshot being deleted - will be fixed by https://github.com/kubernetes-csi/external-snapshotter/pull/1231.

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.31

iPraveenParihar commented:
passed [ci logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.31/detail/mini-e2e_k8s-1.31/227/pipeline/)

@Madhu-1 @Rakshith-R PTAL

Madhu-1 commented on e2e/volumegroupsnapshot.go:
```diff
@@ -176,6 +176,14 @@ func (rvgs *rbdVolumeGroupSnapshot) ValidateResourcesForCreate(vgs *groupsnapapi
+
+	err := waitToRemoveImagesFromTrash(rvgs.framework, defaultRBDPool, deployTimeout)
+	if err != nil {
+		framework.Failf("failed to validate rbd images in pool %s trash: %v", defaultRBDPool, err)
```
return error message

Madhu-1 replied:
its already logged at the caller, lets remove the extra logging from here

nixpanic replied:
I am not sure if `framework.Failf()` really returns, does the test not get aborted when it is called?

Madhu-1 commented on internal/rbd/snapshot.go:
```diff
@@ -162,6 +162,30 @@ func (rbdSnap *rbdSnapshot) ToCSI(ctx context.Context) (*csi.Snapshot, error) {
+	rbdSnap.RbdImageName = rbdVol.RbdImageName
+	err = cleanUpSnapshot(ctx, rbdVol, rbdSnap, rbdVol)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to delete image: %v", err)
```
log the details of what image and snapshot was tried for cleanup for better debugging.

Madhu-1 commented:
Can you please increase the pvc count and put the vgs test in loop of 5 or 10 (as extra commit) to ensure the issue is fixed as we are close to release?

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.29

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.31

iPraveenParihar commented:
undoSnapReservation was missed for `/DeleteVolumeGroupSnapshot`. Now moved `undoSnapReservation()` function to `snapshot.Delete()` so it works for both `/DeleteSnapshot` and `/DeleteVolumeGroupSnapshot`.

cc @Madhu-1 @nixpanic

iPraveenParihar commented:
test volumeGroupSnapshot for both RBD and CephFS passed [here](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e_k8s-1.31/runs/232/nodes/90/log/?start=0). However, it failed due to test timeout of 150m.  

I'll try with 3 PVCs and loop 3 times.

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.31

iPraveenParihar commented:
passed ci [logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.31/detail/mini-e2e_k8s-1.31/233/).

But, OMAP from `csi.groups.default` are not getting removed when VolumeGroupSnapshot Deleted.
There's a bug in `UndoReserve()` for VolumeGroup, it requires volumeGroup `ReqestName` to be passed but `csiID` is being passed here - 
https://github.com/ceph/ceph-csi/blob/a32ba13045f4f433d4c00f568c67209556428a76/internal/rbd/group/util.go#L303-L307

This is not getting captured in e2e as we don't have check for RBD groupSnap OMAP data -
https://github.com/ceph/ceph-csi/blob/a32ba13045f4f433d4c00f568c67209556428a76/e2e/utils.go#L165-L217

I'll add fix for both in this PR.

cc @Madhu-1 @nixpanic @Rakshith-R

iPraveenParihar committed:
rbd: fix volumeGroup UndoReservation

This commit fixes the VolumeGroup UndoReservation
by using the correct RequestName of the VolumeGroup
instead of the volumeGroupHandle.

Signed-off-by: Praveen M <m.praveen@ibm.com>

Madhu-1 commented on internal/rbd/snapshot.go:
```diff
@@ -162,6 +162,38 @@ func (rbdSnap *rbdSnapshot) ToCSI(ctx context.Context) (*csi.Snapshot, error) {
 }
 
+// Delete removes the snapshot from the RBD image and then
+// the RBD image itself.
```
update the comment to mention that it also undo the reservation

Madhu-1 reviewed (commented):
LGTM, small nit

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -1471,27 +1471,9 @@ func (cs *ControllerServer) DeleteSnapshot(
-		return nil, status.Error(codes.Internal, err.Error())
-	}
-	err = undoSnapReservation(ctx, rbdSnap, cr)
+	err = rbdSnap.Delete(ctx)
```
the commit message refers to a cherry-picked commit. Please remove that line, or make sure this PR depends on the PR that adds that commit.

nixpanic commented on internal/rbd/group/util.go:
```diff
@@ -300,7 +309,7 @@ func (cvg *commonVolumeGroup) Delete(ctx context.Context) error {
 	}
 
-	err = j.UndoReservation(ctx, pool, name, csiID)
+	err = j.UndoReservation(ctx, pool, name, reqName)
```
ah, looks like `UndoReservation()` might be improved to return `ErrNotFound` so that this bug could have been caught earlier.

iPraveenParihar replied:
We are ignoring the `ErrObjectNotFound` error, which seems to have been present since the beginning in the volume journal implementation as well.

https://github.com/ceph/ceph-csi/blob/83b54541971b93909a48640a9b4cb84d55e920d6/internal/journal/voljournal.go#L467-L474

Let's address this in different PR? @Madhu-1

nixpanic reviewed (commented):
LGTM, I guess the last two commits should be removed before this gets merged. Marking it `DNM` for the time being.

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.31

iPraveenParihar commented:
passed [ci](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.31/detail/mini-e2e_k8s-1.31/263/pipeline) logs


@Madhu-1 @Rakshith-R PTAL, if its good to merge then I can remove the temp commits before merge.

Rakshith-R commented:
@iPraveenParihar 
there's golangci and go-test failures
https://github.com/ceph/ceph-csi/actions/runs/12347243001/job/34454034537?pr=5000

please fix this
other than this rest of the changes looks good to me.

Madhu-1 reviewed (commented):
LGTM, please drop the extra commit.

iPraveenParihar committed:
rbd: add rbdSnap.Delete() function

This function deletes rbd snap and rbd image
backing k8s snapshot.
The same function is used for deleting
individual snapshots in group snapshot.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar commented:
@Mergifyio queue

Madhu-1 commented:
@mergifyio backport release-v3.13
